include/nccl_ofi_freelist.h

```diff
@@ -20,15 +20,28 @@ extern "C" {
 #include "nccl_ofi_pthread.h"
 
 /*
- * Internal: freelist element structure, only has meaning when the
- * element is in the freelist (as opposed to owned by the user).  Will
- * be the first N bytes of the element buffer when not using memory
- * registration.
+ * Structure describing the registration information for the freelist
+ * item returned by nccl_ofi_freelist_entry_alloc.  The object being
+ * managed by the freelist (such as a control buffer) should contain
+ * this structure starting reginfo_offset bytes from the base of the
+ * structure.  The fields should not be modified by the caller.
  */
-struct nccl_ofi_freelist_elem_t {
-	void *ptr;
-	struct nccl_ofi_freelist_elem_t *next;
+struct nccl_ofi_freelist_reginfo_t {
+	/* offset from the start of the memory registration for the
+	   start of this buffer */
+	size_t base_offset;
```

Comment:
are you sure we need the `base_offset` field in this implementation?

Reply:
We don't need it today, but we would need it to support providers that do not use `FI_MR_VIRT_ADDR`. With this PR, we are still registering memory at a block level, so individual entries will still have a non-zero MR offset from the block.

That was the original intent of the `base_offset` field: https://github.com/aws/aws-ofi-nccl/blob/d1e2fd8e076b34c0296b2354b91a6c9340b60ada/include/nccl_ofi_freelist.h#L166-L169

Reply:
then remove it

Reply:
> then remove it

Ok. But I'm not sure I understand why you want to move further from supporting non-mr-virt-addr mode. I don't think it would be a ton of work, and the TCP provider, which would provide a useful sanity check, is a non-mr-virt-addr provider.

Reply:
because i'm unconvinced this is implemented correctly, and there's no way you tested it.

Reply:
Removed